core/src/test/java/org/apache/iceberg/avro/TestReadProjection.java

```diff
@@ -526,4 +527,216 @@ public void testListOfStructsProjection() throws IOException {
     Assert.assertNull("Should not project y", projectedP2.get("y"));
     Assert.assertEquals("Should project null z", null, projectedP2.get("z"));
   }
+
+  @Test
+  public void testEmptyStructProjection() throws Exception {
+    Schema writeSchema = new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(3, "location", Types.StructType.of(
+            Types.NestedField.required(1, "lat", Types.FloatType.get()),
+            Types.NestedField.required(2, "long", Types.FloatType.get())
+        ))
+    );
+
+    Record record = new Record(AvroSchemaUtil.convert(writeSchema, "table"));
+    record.put("id", 34L);
+    Record location = new Record(
+        AvroSchemaUtil.fromOption(record.getSchema().getField("location").schema()));
+    location.put("lat", 52.995143f);
+    location.put("long", -1.539054f);
+    record.put("location", location);
+
+    Schema emptyStruct = new Schema(
+        Types.NestedField.required(3, "location", Types.StructType.of())
+    );
+
+    Record projected = writeAndRead("empty_proj", writeSchema, emptyStruct, record);
+    Assert.assertNull("Should not project data", projected.get("data"));
+    Record result = (Record) projected.get("location");
```

Comment:
I think we should access the fields that are present by position as well. The expected position for this is 0, so asserting that the same record is returned would be a good check.

Reply:
Wouldn't this be the same as checking if "id" is not projected? Maybe that's a better check? I can add that in too